presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushTopNThroughProject.java

```diff
@@ -0,0 +1,110 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import io.prestosql.matching.Capture;
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.optimizations.SymbolMapper;
+import io.prestosql.sql.planner.plan.Assignments;
+import io.prestosql.sql.planner.plan.FilterNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.ProjectNode;
+import io.prestosql.sql.planner.plan.TableScanNode;
+import io.prestosql.sql.planner.plan.TopNNode;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.SymbolReference;
+
+import java.util.List;
+import java.util.Optional;
+
+import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.sql.planner.plan.Patterns.project;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static io.prestosql.sql.planner.plan.Patterns.topN;
+
+/**
+ * Transforms:
+ * <pre>
+ * - TopN
+ *    - Project (non-identity)
+ *       - Source other than Filter(TableScan) / TableScan
+ * </pre>
+ * Into:
+ * <pre>
+ * - Project
+ *    - TopN
+ *       - Source
+ * </pre>
+ */
+public final class PushTopNThroughProject
+        implements Rule<TopNNode>
+{
+    private static final Capture<ProjectNode> PROJECT_CHILD = newCapture();
+
+    private static final Pattern<TopNNode> PATTERN =
+            topN()
+                    .with(source().matching(
+                            project()
+                                    .matching(projectNode -> !projectNode.isIdentity())
+                                    .capturedAs(PROJECT_CHILD)
+                                    .with(source().matching(node -> !(node instanceof TableScanNode)))));
+
+    @Override
+    public Pattern<TopNNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(TopNNode parent, Captures captures, Context context)
+    {
+        ProjectNode projectNode = captures.get(PROJECT_CHILD);
+
+        PlanNode projectSource = context.getLookup().resolve(projectNode.getSource());
+        if (projectSource instanceof FilterNode) {
+            PlanNode filterSource = context.getLookup().resolve(((FilterNode) projectSource).getSource());
+            if (filterSource instanceof TableScanNode) {
+                return Result.empty();
```

Comment:
Why do we limit this case? I would think this rule should not care about whether there's a TableScan or Filter involved.

Reply:
also, there's another limitation in the `PATTERN`, which says that there cannot be a `TableScanNode` as a source of the Projection. Probably both limitations should go to `apply()` for symmetry.
> Why do we limit this case?

I was about to ask @sopel39 and @kokosing about that, because it was their request. I want to add some explanatory comment.

Reply:
@martint if you push topN below projection that is above filter or table scan then such projection won't be merged as a single `PageProcessor` with filter or/and table scan.

If such topN node is `PARTIAL` and we have a lot of splits then it could potentially be more expensive to split projection from filter. What do you think?

Reply:
> also, there's another limitation in the PATTERN, which says that there cannot be a TableScanNode as a source of the Projection. Probably both limitations should go to apply() for symmetry.

It's fine to have different but complementary constraints in the pattern vs in apply. Sometimes there are complex cases that cannot be easily described in terms of patterns. But if they can, there's better opportunity for the pattern matcher to do it more efficiently.

>  I want to add some explanatory comment.

Yes, please do! My general rule of thumb is "if I were to look at some code I wrote a few months down the road, would I be able to figure out what it's doing or the motivation?". If not, a comment is warranted. In this case, it's not entirely obvious why this pattern is excluded, so let's add an explanation.